File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
from heapq import heappush, heappop, heapify
from collections import defaultdict
def solution(operations):
    
    max_q = []
    min_q = []
    record = defaultdict(int)
    
    for operation in operations:
        command, num = operation.split()
        
        if command == 'I':
            heappush(min_q, int(num))
            heappush(max_q, -int(num))
            record[int(num)] += 1
        elif num == '1': # 최댓값 삭제
            while max_q:
                remove = heappop(max_q)
                if record[-remove] > 0:
                    record[-remove] -= 1
                    break
        else: # num == '-1', 최솟값 삭제 
            while min_q:
                remove = heappop(min_q)
                if record[remove] > 0:
                    record[remove] -= 1
                    break

    mx, mn = 0, 0
    
    while max_q:
            mx = heappop(max_q)
            if record[-mx] > 0:
                break
            mx = 0
    while min_q:
            mn = heappop(min_q)
            if record[mn] > 0:
                break
            mn = 0
            
    return [-mx, mn]

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*;

class Solution { 
    private int remove(PriorityQueue<Integer> pq, HashMap<Integer, Integer> map) {
        while(!pq.isEmpty()) {
            int remove = pq.poll();
            if (map.get(remove) > 0) {
                map.put(remove, map.get(remove) - 1);
                return remove;
            }
        }
        return 0;
    }
    
    public int[] solution(String[] operations) {
        PriorityQueue<Integer> minQ = new PriorityQueue<>();
        PriorityQueue<Integer> maxQ = new PriorityQueue<>((a, b) -> b - a);
        HashMap<Integer, Integer> countMap = new HashMap<>();
        
        for(String op : operations) {
            String command = op.split(" ")[0];
            int data = Integer.parseInt(op.split(" ")[1]);
            
            if(command.equals("I")) {
                countMap.put(data, countMap.getOrDefault(data, 0) + 1);
                minQ.add(data); maxQ.add(data);
            } else {
                if(data == 1) remove(maxQ, countMap);
                else remove(minQ, countMap);
            }
        }
        
        int max = 0; int min = 0;
        while(!maxQ.isEmpty() && countMap.get(maxQ.peek()) <= 0) maxQ.poll();
        if(!maxQ.isEmpty()) max = maxQ.poll();
        while(!minQ.isEmpty() && countMap.get(minQ.peek()) <= 0) minQ.poll();
        if(!minQ.isEmpty()) min = minQ.poll();
        
        return new int[]{max, min};
    }
}
